File: core/utils/observability.ts

export type ErrorCategory = 'AUTH' | 'NETWORK' | 'BUILD' | 'VALIDATION' | 'DEPENDENCY';

export interface CorrelationContext {
  requestId: string;
  correlationId: string;
  authContext: string;
}

interface StructuredLog {
  timestamp?: string;
  component: string;
  operation: string;
  auth_context: string;
  request_id: string;
  correlation_id: string;
  error_code?: string;
  level?: 'INFO' | 'WARN' | 'ERROR';
  message?: string;
  details?: Record<string, unknown>;
}

export class StructuredError extends Error {
  constructor(
    public readonly category: ErrorCategory,
    message: string,
    public readonly status = 500,
    public readonly details?: Record<string, unknown>
  ) {
    super(message);
    this.name = 'StructuredError';
  }

  toResponseBody(context: CorrelationContext) {
    return {
      success: false,
      error: this.message,
      category: this.category,
      requestId: context.requestId,
      correlationId: context.correlationId,
      details: this.details,
    };
  }
}

const makeId = () => {
  if (typeof window !== 'undefined' && window.crypto?.randomUUID) {
    return window.crypto.randomUUID();
  }
  return Math.random().toString(36).substring(2) + Date.now().toString(36);
};

export function resolveCorrelationContext(
  headers: Headers,
  authHeader?: string | null
): CorrelationContext {
  const requestId = headers.get('x-request-id') || makeId();
  const correlationId = headers.get('x-correlation-id') || requestId;
  const authContext = authHeader?.startsWith('Bearer ') ? 'BEARER' : 'ANONYMOUS';

  return { requestId, correlationId, authContext };
}

export function logStructured(entry: StructuredLog): void {
  const payload: StructuredLog = {
    ...entry,
    timestamp: entry.timestamp || new Date().toISOString(),
    level: entry.level || 'INFO',
  };

  const line = JSON.stringify(payload);
  if (payload.level === 'ERROR') {
    console.error(line);
    return;
  }

  if (payload.level === 'WARN') {
    console.warn(line);
    return;
  }

  console.log(line);
}

export function mapUnknownError(error: unknown): StructuredError {
  if (error instanceof StructuredError) {
    return error;
  }

  if (error instanceof Error) {
    return new StructuredError('DEPENDENCY', error.message, 500);
  }

  return new StructuredError('DEPENDENCY', 'Unknown server error', 500);
}

export function withCorrelationHeaders(
  init: ResponseInit | undefined,
  context: CorrelationContext
): ResponseInit {
  return {
    ...init,
    headers: {
      ...(init?.headers || {}),
      'x-request-id': context.requestId,
      'x-correlation-id': context.correlationId,
    },
  };
}

File: sidecar/sovereign-engine/pkg/server/observability.go
package server

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"time"

	"google.golang.org/grpc/metadata"
)

type ErrorCategory string

const (
	ErrorAuth       ErrorCategory = "AUTH"
	ErrorNetwork    ErrorCategory = "NETWORK"
	ErrorBuild      ErrorCategory = "BUILD"
	ErrorValidation ErrorCategory = "VALIDATION"
	ErrorDependency ErrorCategory = "DEPENDENCY"
)

type logEvent struct {
	Timestamp     string `json:"timestamp"`
	Component     string `json:"component"`
	Operation     string `json:"operation"`
	AuthContext   string `json:"auth_context"`
	RequestID     string `json:"request_id"`
	CorrelationID string `json:"correlation_id"`
	ErrorCode     string `json:"error_code,omitempty"`
	Message       string `json:"message,omitempty"`
	Level         string `json:"level,omitempty"`
}

func extractTraceContext(ctx context.Context) (requestID string, correlationID string, authContext string) {
	requestID = fmt.Sprintf("go-request-%d", time.Now().UTC().UnixNano())
	correlationID = requestID
	authContext = "ANONYMOUS"

	if md, ok := metadata.FromIncomingContext(ctx); ok {
		if v := md.Get("x-request-id"); len(v) > 0 && v[0] != "" {
			requestID = v[0]
		}
		if v := md.Get("x-correlation-id"); len(v) > 0 && v[0] != "" {
			correlationID = v[0]
		} else {
			correlationID = requestID
		}
		if v := md.Get("x-sovereign-token"); len(v) > 0 && v[0] != "" {
			authContext = "SOVEREIGN_TOKEN"
		}
	}

	return requestID, correlationID, authContext
}

func logStructured(ctx context.Context, operation string, level string, message string, errorCode ErrorCategory) {
	requestID, correlationID, authContext := extractTraceContext(ctx)
	event := logEvent{
		Timestamp:     time.Now().UTC().Format(time.RFC3339Nano),
		Component:     "GO_SIDECAR",
		Operation:     operation,
		AuthContext:   authContext,
		RequestID:     requestID,
		CorrelationID: correlationID,
		Level:         level,
		Message:       message,
	}
	if errorCode != "" {
		event.ErrorCode = string(errorCode)
	}

	line, err := json.Marshal(event)
	if err != nil {
		log.Printf("{\"timestamp\":\"%s\",\"component\":\"GO_SIDECAR\",\"operation\":\"%s\",\"level\":\"ERROR\",\"message\":\"log serialization failure\",\"error_code\":\"DEPENDENCY\"}", time.Now().UTC().Format(time.RFC3339Nano), operation)
		return
	}

	log.Println(string(line))
}
